### ISSUE TITLE
Java Heap Space 1.3.5 on Arch 5.7.4

### ISSUE THREAD

**Issue by Robert-L-Turner** [bisq-network/bisq#4363]
### Description

Java Heap Space - Out of Memory Error after leaving application open for extended period

#### Version

1.3.5

### Steps to reproduce

Leave app open for 12+ hours.  Can see a consistent reduction in available memory.  Restart application clears issue.  

### Screenshots

![image](https://user-images.githubusercontent.com/37597024/87016356-2fbea300-c201-11ea-9bad-dbadb01ef164.png)

#### Device or machine

Arch 5.7.4

**Comment by cd2357** [comment]
How much RAM total did the system have?

Also, was this a VM?

**Comment by Robert-L-Turner** [comment]
8gb of ram, not a virtual machine.

**Comment by re42** [comment]
I have been getting the same error once or twice after a prolonged period of time since version 1.3.5.  I also got it now again in 1.3.7.  I was just too lazy to report it previously.
Honestly, there is something wrong here.
and yes i have 32GB RAM, so its not a memory problem.  Not a VM.

Xubuntu 18.04

Below log
```
Aug-10 15:46:14.296 [JavaFX Application Thread] INFO  b.d.app.BisqApp: System load: Memory (MB): 1004 / No. of threads: 66 
Aug-10 15:47:20.973 [JavaFX Application Thread] INFO  b.c.p.f.FeeService: BTC tx fee: txFeePerByte=120 
Aug-10 15:50:04.503 [Save-file-task-3827] ERROR b.c.s.FileManager: Error in saveToFile toProtoMessage: TradeStatistics2Store, /home/user/.local/share/Bisq/btc_mainnet/db/TradeStatistics2Store 
Aug-10 15:50:04.521 [Save-file-task-3827] ERROR b.c.s.FileManager: Error at saveToFile, storageFile=/home/user/.local/share/Bisq/btc_mainnet/db/TradeStatistics2Store java.lang.RuntimeException: java.lang.OutOfMemoryError: Java heap space
	at bisq.common.storage.FileManager.saveToFile(FileManager.java:200)
	at bisq.common.storage.FileManager.lambda$new$0(FileManager.java:81)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at com.google.protobuf.AbstractMessageLite.toByteArray(AbstractMessageLite.java:69)
	at bisq.common.storage.FileManager.saveToFile(FileManager.java:195)
	at bisq.common.storage.FileManager.lambda$new$0(FileManager.java:81)
	at bisq.common.storage.FileManager$$Lambda$202/733899724.call(Unknown Source)
	... 5 common frames omitted

Aug-10 15:51:37.801 [Save-file-task-6092] ERROR b.c.s.FileManager: Error in saveToFile toProtoMessage: TradeStatistics2Store, /home/user/.local/share/Bisq/btc_mainnet/db/TradeStatistics2Store 
Aug-10 15:51:37.802 [Save-file-task-6092] ERROR b.c.s.FileManager: Error at saveToFile, storageFile=/home/user/.local/share/Bisq/btc_mainnet/db/TradeStatistics2Store java.lang.RuntimeException: java.lang.OutOfMemoryError: Java heap space
	at bisq.common.storage.FileManager.saveToFile(FileManager.java:200)
	at bisq.common.storage.FileManager.lambda$new$0(FileManager.java:81)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at com.google.protobuf.AbstractMessageLite.toByteArray(AbstractMessageLite.java:69)
	at bisq.common.storage.FileManager.saveToFile(FileManager.java:195)
	at bisq.common.storage.FileManager.lambda$new$0(FileManager.java:81)
	at bisq.common.storage.FileManager$$Lambda$202/733899724.call(Unknown Source)
	... 5 common frames omitted

Aug-10 15:52:19.586 [JavaFX Application Thread] INFO  b.c.p.f.FeeService: BTC tx fee: txFeePerByte=121 
Aug-10 15:54:13.494 [Save-file-task-1734] ERROR b.c.s.FileManager: Error in saveToFile toProtoMessage: TradeStatistics2Store, /home/user/.local/share/Bisq/btc_mainnet/db/TradeStatistics2Store 
Aug-10 15:54:13.494 [PeerGroup Thread] ERROR o.b.c.PeerGroup: Error in speed calculator java.lang.OutOfMemoryError: Java heap space

Aug-10 15:54:13.495 [Save-file-task-1734] ERROR b.c.s.FileManager: Error at saveToFile, storageFile=/home/user/.local/share/Bisq/btc_mainnet/db/TradeStatistics2Store java.lang.RuntimeException: java.lang.OutOfMemoryError: Java heap space
	at bisq.common.storage.FileManager.saveToFile(FileManager.java:200)
	at bisq.common.storage.FileManager.lambda$new$0(FileManager.java:81)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.OutOfMemoryError: Java heap space

Aug-10 15:55:13.810 [JavaFX Application Thread] ERROR b.c.s.CommonSetup: Uncaught Exception from thread JavaFX Application Thread 
Aug-10 15:55:13.810 [JavaFX Application Thread] ERROR b.c.s.CommonSetup: throwableMessage= Java heap space 
Aug-10 15:55:13.811 [JavaFX Application Thread] ERROR b.c.s.CommonSetup: throwableClass= class java.lang.OutOfMemoryError 
Aug-10 15:55:15.391 [Save-file-task-1176] ERROR b.c.s.FileManager: Error in saveToFile toProtoMessage: TradeStatistics2Store, /home/user/.local/share/Bisq/btc_mainnet/db/TradeStatistics2Store 
Aug-10 15:55:15.392 [Save-file-task-1176] ERROR b.c.s.FileManager: Error at saveToFile, storageFile=/home/user/.local/share/Bisq/btc_mainnet/db/TradeStatistics2Store java.lang.RuntimeException: java.lang.OutOfMemoryError: Java heap space
	at bisq.common.storage.FileManager.saveToFile(FileManager.java:200)
	at bisq.common.storage.FileManager.lambda$new$0(FileManager.java:81)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.OutOfMemoryError: Java heap space

Aug-10 15:55:15.395 [JavaFX Application Thread] ERROR b.c.s.CommonSetup: Stack trace:
java.lang.OutOfMemoryError: Java heap space
 
Aug-10 15:55:15.910 [JavaFX Application Thread] INFO  b.n.p.p.PeerManager: onDisconnect called: nodeAddress=Optional[wizbisqzd7ku25di7p2ztsajioabihlnyp5lq5av66tmu7do2dke2tid.onion:9999], closeConnectionReason=CloseConnectionReason{sendCloseMessage=false, isIntended=false} NO_PROTO_BUFFER_ENV 
Aug-10 15:55:15.912 [JavaFX Application Thread] INFO  b.n.p.p.p.PeerExchangeManager: onDisconnect closeConnectionReason=CloseConnectionReason{sendCloseMessage=false, isIntended=false} NO_PROTO_BUFFER_ENV, nodeAddressOpt=Optional[wizbisqzd7ku25di7p2ztsajioabihlnyp5lq5av66tmu7do2dke2tid.onion:9999] 
Aug-10 15:55:29.654 [JavaFX Application Thread] INFO  b.n.p.p.PeerManager: Num of latestLivePeers=1 
Aug-10 15:55:43.005 [JavaFX Application Thread] INFO  b.n.p.p.PeerManager: We have 8 connections open. Our limit is 12 
Aug-10 15:56:04.965 [Thread-24] INFO  b.c.o.OpenOfferManager: Remove open offers at shutDown. Number of open offers: 3 
Aug-10 15:56:04.973 [Thread-24] INFO  b.c.a.AvoidStandbyModeService: Stopped 
Aug-10 15:56:06.098 [JavaFX Application Thread] INFO  b.c.a.BisqExecutable: OpenOfferManager shutdown completed 
Aug-10 15:56:06.100 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown 8 connections 
Aug-10 15:56:06.105 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown of node Optional[jnws6zrnztsa6qpi.onion:8000] completed 
Aug-10 15:56:06.105 [BlockingClient network thread for 4nnuyxm5k5tlyjq3.onion:8333] INFO  o.b.c.PeerGroup: [4nnuyxm5k5tlyjq3.onion]:8333: Peer died      (8 connected, 0 pending, 9 max) 
Aug-10 15:56:06.105 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 1000 msec before next connect attempt to [lgkvbvro67jomosw.onion]:8333 
Aug-10 15:56:06.108 [BlockingClient network thread for jiuuuislm7ooesic.onion:8333] INFO  o.b.c.PeerGroup: [jiuuuislm7ooesic.onion]:8333: Peer died      (7 connected, 0 pending, 9 max) 
Aug-10 15:56:06.108 [BlockingClient network thread for z6mbqq7llxlrn4kq.onion:8333] INFO  o.b.c.PeerGroup: [z6mbqq7llxlrn4kq.onion]:8333: Peer died      (6 connected, 0 pending, 9 max) 
Aug-10 15:56:06.108 [BlockingClient network thread for 2pj2o2mrawj7yotg.onion:8333] INFO  o.b.c.PeerGroup: [2pj2o2mrawj7yotg.onion]:8333: Peer died      (5 connected, 0 pending, 9 max) 
Aug-10 15:56:06.108 [BlockingClient network thread for 2pj2o2mrawj7yotg.onion:8333] INFO  o.b.c.PeerGroup: Download peer died. Picking a new one. 
Aug-10 15:56:06.108 [BlockingClient network thread for 2pj2o2mrawj7yotg.onion:8333] INFO  o.b.c.PeerGroup: Unsetting download peer: [2pj2o2mrawj7yotg.onion]:8333 
Aug-10 15:56:06.108 [BlockingClient network thread for 2pj2o2mrawj7yotg.onion:8333] INFO  o.b.c.PeerGroup: Setting download peer: [orsy2v63ecrmdj55.onion]:8333 
Aug-10 15:56:06.108 [BlockingClient network thread for 2pj2o2mrawj7yotg.onion:8333] INFO  o.b.c.l.DownloadProgressTracker: Chain download switched to [orsy2v63ecrmdj55.onion]:8333 
Aug-10 15:56:06.109 [BlockingClient network thread for 2pj2o2mrawj7yotg.onion:8333] WARN  o.b.n.BlockingClient: Attempted to write to a closed socket. 
Aug-10 15:56:06.109 [BlockingClient network thread for orsy2v63ecrmdj55.onion:8333] INFO  o.b.c.PeerGroup: [orsy2v63ecrmdj55.onion]:8333: Peer died      (4 connected, 0 pending, 9 max) 
Aug-10 15:56:06.109 [BlockingClient network thread for orsy2v63ecrmdj55.onion:8333] INFO  o.b.c.PeerGroup: Download peer died. Picking a new one. 
Aug-10 15:56:06.113 [BlockingClient network thread for orsy2v63ecrmdj55.onion:8333] INFO  o.b.c.PeerGroup: Unsetting download peer: [orsy2v63ecrmdj55.onion]:8333 
Aug-10 15:56:06.113 [BlockingClient network thread for orsy2v63ecrmdj55.onion:8333] INFO  o.b.c.PeerGroup: Setting download peer: [fz6nsij6jiyuwlsc.onion]:8333 
Aug-10 15:56:06.113 [BlockingClient network thread for orsy2v63ecrmdj55.onion:8333] INFO  o.b.c.l.DownloadProgressTracker: Chain download switched to [fz6nsij6jiyuwlsc.onion]:8333 
Aug-10 15:56:06.113 [BlockingClient network thread for orsy2v63ecrmdj55.onion:8333] WARN  o.b.n.BlockingClient: Attempted to write to a closed socket. 
Aug-10 15:56:06.113 [BlockingClient network thread for fz6nsij6jiyuwlsc.onion:8333] INFO  o.b.c.PeerGroup: [fz6nsij6jiyuwlsc.onion]:8333: Peer died      (3 connected, 0 pending, 9 max) 
Aug-10 15:56:06.113 [BlockingClient network thread for fz6nsij6jiyuwlsc.onion:8333] INFO  o.b.c.PeerGroup: Download peer died. Picking a new one. 
Aug-10 15:56:06.113 [BlockingClient network thread for fz6nsij6jiyuwlsc.onion:8333] INFO  o.b.c.PeerGroup: Unsetting download peer: [fz6nsij6jiyuwlsc.onion]:8333 
Aug-10 15:56:06.113 [BlockingClient network thread for fz6nsij6jiyuwlsc.onion:8333] INFO  o.b.c.PeerGroup: Setting download peer: [i3a5xtzfm4xwtybd.onion]:8333 
Aug-10 15:56:06.113 [BlockingClient network thread for fz6nsij6jiyuwlsc.onion:8333] INFO  o.b.c.l.DownloadProgressTracker: Chain download switched to [i3a5xtzfm4xwtybd.onion]:8333 
Aug-10 15:56:06.114 [BlockingClient network thread for fz6nsij6jiyuwlsc.onion:8333] WARN  o.b.n.BlockingClient: Attempted to write to a closed socket. 
Aug-10 15:56:06.114 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 7599 msec before next connect attempt to [mbm6ffx6j5ygi2ck.onion]:8333 
Aug-10 15:56:06.114 [BlockingClient network thread for 3xucqntxp5ddoaz5.onion:8333] INFO  o.b.c.PeerGroup: [3xucqntxp5ddoaz5.onion]:8333: Peer died      (2 connected, 0 pending, 9 max) 
Aug-10 15:56:06.114 [BlockingClient network thread for 3r44ddzjitznyahw.onion:8333] INFO  o.b.c.PeerGroup: [3r44ddzjitznyahw.onion]:8333: Peer died      (1 connected, 0 pending, 9 max) 
Aug-10 15:56:06.114 [BlockingClient network thread for i3a5xtzfm4xwtybd.onion:8333] INFO  o.b.c.PeerGroup: [i3a5xtzfm4xwtybd.onion]:8333: Peer died      (0 connected, 0 pending, 9 max) 
Aug-10 15:56:06.114 [BlockingClient network thread for i3a5xtzfm4xwtybd.onion:8333] INFO  o.b.c.PeerGroup: Download peer died. Picking a new one. 
Aug-10 15:56:06.114 [BlockingClient network thread for i3a5xtzfm4xwtybd.onion:8333] INFO  o.b.c.PeerGroup: Unsetting download peer: [i3a5xtzfm4xwtybd.onion]:8333 
Aug-10 15:56:06.114 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 10000 msec before next connect attempt to [lgkvbvro67jomosw.onion]:8333 
Aug-10 15:56:06.114 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 10000 msec before next connect attempt to [sjyzmwwu6diiit3r.onion]:8333 
Aug-10 15:56:06.114 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 10000 msec before next connect attempt to [lgkvbvro67jomosw.onion]:8333 
Aug-10 15:56:06.115 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 10000 msec before next connect attempt to [sjyzmwwu6diiit3r.onion]:8333 
Aug-10 15:56:06.115 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 9999 msec before next connect attempt to [lgkvbvro67jomosw.onion]:8333 
Aug-10 15:56:06.115 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 9999 msec before next connect attempt to [sjyzmwwu6diiit3r.onion]:8333 
Aug-10 15:56:06.115 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Waiting 9999 msec before next connect attempt to [lgkvbvro67jomosw.onion]:8333 
Aug-10 15:56:06.303 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown of node Optional[swprinmuqadd4pnybcc263dw2bjrrwrt4ivutqaqtb2g6g4d6hybgzyd.onion:9999] completed 
Aug-10 15:56:06.304 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown of node Optional[6za6f74affekizj5.onion:9999] completed 
Aug-10 15:56:06.304 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown of node Optional[sn2bisqad7ncazupgbd3dcedqh5ptirgwofw63djwpdtftwhddo75oid.onion:8000] completed 
Aug-10 15:56:06.304 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown of node Optional[2pcrnibhco3gzwtu.onion:9999] completed 
Aug-10 15:56:06.304 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown of node Optional[xgmtwcsohjc3ezpo.onion:9999] completed 
Aug-10 15:56:06.304 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown of node Optional[7c7ff3amiswoc6og.onion:9999] completed 
Aug-10 15:56:06.304 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown of node Optional[kwz7jtaggyo5xazb.onion:9999] completed 
Aug-10 15:56:06.304 [JavaFX Application Thread] INFO  b.n.p.n.NetworkNode: Shutdown completed with all connections closed 
Aug-10 15:56:06.305 [JavaFX Application Thread] INFO  b.c.a.BisqExecutable: P2PService shutdown completed 
Aug-10 15:56:06.306 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Stopping ... 
Aug-10 15:56:06.306 [ STOPPING] INFO  o.b.c.PeerGroup: Awaiting PeerGroup shutdown ... 
Aug-10 15:56:06.306 [PeerGroup Thread] INFO  o.b.c.PeerGroup: Stopped. 
Aug-10 15:56:11.306 [JavaFX Application Thread] INFO  b.c.a.BisqExecutable: WalletsSetup shutdown completed 
Aug-10 15:56:11.306 [JavaFX Application Thread] INFO  b.c.a.BisqExecutable: Graceful shutdown completed. Exiting now. 
```

After i saw the error, i just closed bisq.

**Comment by re42** [comment]
Hi, @ManfredKarrer can you please take a look at this? bisq is unusable. this keeps happening.

**Comment by freimair** [comment]
in case you guys followed the AUR install:

There is a `bisq-desktop` in `/opt/bisq/`. In the file, there is a line containing `MaxRAM=4g`. Change that to 8g for example but only if you have more than 8g of RAM available.

We are trying to reduce the amount of memory consumed by Bisq but it is mostly due to memory leaks in the JavaFX GUI library we use and so we have no direct way of fixing that. Stay tunes, we are on our way to a better JavaFX version. Sorry for the inconvenience.

**Comment by re42** [comment]
@freimair this issue is NOT specific to Arch, the title should be changed, as i reported above that i am having the same issue in Xubuntu 18.04 using the .deb package.

**Comment by freimair** [comment]
all linux packages are shipped with `MaxRAM=4g`. Finding the `bisq-desktop` executable and changing the value works there as well.

**Comment by re42** [comment]
@freimair thank you, but my `bisq-desktop` executable does not contain `MaxRAM=4g`
Should i just add it?

PS: v1.3.9 also has the same issue, since it crashed for me yesterday.

**Comment by cd2357** [comment]
> ... i am having the same issue in Xubuntu 18.04 using the .deb package.

@re42 In case you still have the Xubuntu installation, you please paste the contents of `/opt/Bisq/app/Bisq.cfg`?

**Comment by re42** [comment]
> > ... i am having the same issue in Xubuntu 18.04 using the .deb package.
> 
> @re42 In case you still have the Xubuntu installation, you please paste the contents of `/opt/Bisq/app/Bisq.cfg`?

Ah I see the MaxRAM option in _that_ file, should i change it there?

contents:
```
[Application]
app.name=Bisq
app.version=1.3.9
app.preferences.id=bisq/desktop/app
app.runtime=$APPDIR/runtime
app.identifier=bisq.desktop.app
app.classpath=
app.application.instance=multiple
app.mainclass=bisq/desktop/app/BisqAppMain
app.mainjar=desktop-1.3.9-all.jar
packager.java.version=10.0.2

[JVMOptions]
-Xss1280k
-XX:MaxRAM=4g
-Djava.net.preferIPv4Stack=true

[JVMUserOptions]

[ArgOptions]
```

**Comment by cd2357** [comment]
Yes, but please make a backup of your Bisq data directory first. See https://bisq.wiki/Backing_up_application_data

Technically, 4g should be more than enough, but maybe you have lots of (open/closed) trades, transactions, messages, tickets in mediation or arbitration, etc which may have smth to do with used memory.

After you backup the data directory, can you tell us what is the size of the resulting backup?

Also: did you see this error since you started using Bisq (i.e. a fresh installation)? Or was it fine in the beginning, but as you used it more, it sometimes gave this heap error?

**Comment by re42** [comment]
ok, ive backed up and its about 400MB (i have a quite a few 10mb log files and and file called: prevent-app-nap-silent-sound.aiff which is about 40MB.)

> Also: did you see this error since you started using Bisq (i.e. a fresh installation)? Or was it fine in the beginning, but as you used it more, it sometimes gave this heap error?

No it wasnt a fresh installation, the error only started happening since after the update to v1.3.5, before that i would leave bisq on for hours and days and it would be fine.
I am almost 100% sure that something happened in v1.3.5 which is currently still the issue in the versions after that.

@freimair said that there are memory leaks in the JavaFX GUI library, is this something that is on/off with different users using bisq?  Did something change in v1.3.5 that would have caused memory leaks?

**Comment by ripcurlx** [comment]
> @freimair said that there are memory leaks in the JavaFX GUI library, is this something that is on/off with different users using bisq? Did something change in v1.3.5 that would have caused memory leaks?

Not that I'm aware off. There is good progress in https://github.com/bisq-network/bisq/pull/4242 which would fix a lot JVM and JavaFX related memory issues (hopefully).

**Comment by cd2357** [comment]
Should be fixed in the most recent release (v1.6.4) which brings several UI performance improvements and generally reduces system resource consumption. Please try it out and let us know if it's still an issue for you.
